Store artifacts durring build

btashton opened the pull request:
## Summary
Setting up some plumbing for being able to store the build artifacts for processing afterwards for things like testing and metrics.

A new option `-A` is added to `tools/testbuild.sh` that will take the created build executable and store it in a folder for the config that generated it under `$ARTIFACTDIR` which can be set via an environment variable or defaulted to `$(TOPDIR)/buildartifacts`

This is also helpful for local testing because you can now run `tools/testbuild.sh -A sim.dat` and have all of the simulation targets generated without having to rebuild along the way.

A zip is created for the artifacts in each run. ~~Unfortunately GitHub Actions does not provides a way capture all of the runs data together so it is a zip for every `.dat` file we build against.~~ this was wrong it helpfully merges them it the name given to the artifact is the same  in this case `builds`. So in this case 100+ elf files are sorted in a tree based on config.

## Impact
Artifacts should be able to be attached to the CI run.

## Testing
CI

btashton commented:
One thing I am not quite happy with here is we do not expose other artifacts that might be expected for a build.  Perhaps later we could provide a make target that exposes all of the them and not just the executable.

btashton commented:
This actually worked better than I thought. The artifacts got all merged into one archive when they are uploaded
![Screenshot_20201025-235009.png](https://user-images.githubusercontent.com/173245/97142026-1b090b00-171d-11eb-9925-5a977814f145.png)

btashton commented:
@xiaoxiang781216 any thoughts on this? It would really speed up my testing.

xiaoxiang781216 commented on tools/testbuild.sh:
```diff
@@ -238,6 +246,11 @@ function configure {
+  if [ ${SAVEARTIFACTS} -eq 1 ]; then
+    artifactconfigdir=$ARTIFACTDIR/$(echo $config | sed "s/:/\//")/
+    mkdir -p $artifactconfigdir
+    cp $(${MAKE} execname) $artifactconfigdir
```
should we copy more file here? e.g. nuttx.bin, System.map.. which is useful to debug the problem.

btashton replied:
Yeah but I don't know how to get those. If we can figure out how to echo, any recommendations? I figured start with the executable and refine going forward.

Ouss4 replied:
For most, we have hard-coded names that has never changed.

btashton replied:
I'll generate a manifest file in the Makefile.(unix/win) that can be used for this.  Might need to extend the boards a bit so we know about things like uboot images.

xiaoxiang781216 commented on tools/testbuild.sh:
```diff
@@ -62,6 +66,7 @@ function showusage {
   echo "  -a <appsdir> provides the relative path to the apps/ directory.  Default ../apps"
   echo "  -t <topdir> provides the absolute path to top nuttx/ directory.  Default $PWD/../nuttx"
   echo "  -p only print the list of configs without running any builds"
+  echo "  -A store the build executable artifact in ARTIFACTDIR (defaults to $PWD/../buildartifacts"
```
let's remove $PWD here and line 67 since no other place use it in the whole script

btashton replied:
I came make it relative to topdir. Does that work?

xiaoxiang781216 commented on Makefile:
```diff
@@ -45,3 +45,7 @@ else
+
+.PHONY: execname
+execname:
+	@echo $(BIN)
```
why not copy the binary directly? it's strange to print the binary name in Makefile

btashton replied:
I want a target that allows other scripting to use the files (like out CI system). I don't want the makefile to actually do anything with them, right now there is no way to get this info. We could generate an manifest file if you prefer.

btashton replied:
Manifest file is now generated by the build that can be used.

Ouss4 replied:
You don't need this anymore, right?

xiaoxiang781216 replied:
@btashton let's revert this change?

btashton replied:
Must have slipped in when I was resolving merge conflicts.  I have removed and rebased to fix the other conflict.

xiaoxiang781216 commented on .github/workflows/build.yml:
```diff
@@ -161,8 +161,13 @@ jobs:
             mkdir $CCACHE_DIR
             cd sources/testing
-            ./cibuild.sh -c testlist/${{matrix.boards}}.dat
+            export ARTIFACTDIR=`pwd`/../../buildartifacts
```
Remove all `pwd`/ since other place use the relative path.

btashton replied:
The exported variable needs an absolute path here or we risk dumping it in a random place.

btashton commented:
@xiaoxiang781216 Let me know if this new version addresses your concerns:

The output now looks like this:
```
../buildartifacts/ 
├── beaglebone-black
│   ├── lcd
│   │   ├── nuttx
│   │   ├── nuttx.bin
│   │   ├── nuttx.hex
│   │   └── System.map
│   └── nsh
│       ├── nuttx
│       ├── nuttx.bin
│       ├── nuttx.hex
│       └── System.map
├── c5471evm
│   ├── httpd
│   │   ├── nuttx
│   │   └── System.map
│   ├── nettest
│   │   ├── nuttx
│   │   └── System.map
│   └── nsh
│       ├── nuttx
│       └── System.map
```

I am also generating a nuttx.manifest file rather than using make to echo out the target.

xiaoxiang781216 commented on tools/Makefile.win:
```diff
@@ -392,17 +392,22 @@ ifeq ($(CONFIG_BUILD_2PASS),y)
 endif
 	$(Q) $(MAKE) -C $(ARCH_SRC) EXTRA_OBJS="$(EXTRA_OBJS)" LINKLIBS="$(LINKLIBS)" EXTRAFLAGS="$(KDEFINE) $(EXTRAFLAGS)" $(BIN)
+	echo $(BIN) > $(NUTTXNAME).manifest
+	printf '%s\n' *.map >> $(NUTTXNAME).manifest
```
can we use echo "*.map"?

btashton replied:
echo will put them all on the same line so it would look like this:

```
nuttx
nuttx.hex
System.map User.map
```

I think printf is the best option

xiaoxiang781216 reviewed (approved):
LGTM.

btashton commented:
Looks like macOS `xargs` does not support the `-a`
```
xargs: illegal option -- a
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
             [-L number] [-n number [-x]] [-P maxprocs] [-s size]
             [utility [argument ...]]
```
I'll fix this in a follow up PR. It's does not break the build we just don't capture the macOS artifacts.
